amm/contracts/router_v2/pair.rs

```diff
@@ -0,0 +1,226 @@
+use crate::utils::*;
+use amm_helpers::{ensure, math::casted_mul};
+use ink::{
+    codegen::TraitCallBuilder,
+    contract_ref,
+    env::{account_id, caller, transfer, transferred_value, DefaultEnvironment as Env},
+    prelude::string::String,
+    primitives::AccountId,
+};
+use traits::{Balance, MathError, Pair as PairTrait, RouterV2Error};
+#[derive(scale::Decode, scale::Encode)]
+#[cfg_attr(
+    feature = "std",
+    derive(scale_info::TypeInfo, ink::storage::traits::StorageLayout)
+)]
+pub struct Pair(AccountId, u8);
+
+impl Pair {
+    pub fn new(pair_id: AccountId) -> Self {
+        let mut pair = Self(pair_id, 0);
+        pair.1 = pair.contract_ref().get_fee();
+        pair
+    }
+    /// Makes a cross-contract call to fetch the Pair reserves.
+    /// Returns reserves `(reserve_0, reserve_1)` in order of `token_0` and `token_1`
+    pub fn get_reserves(&self, token_0: AccountId, token_1: AccountId) -> (u128, u128) {
```

Comment:
Yes, that was my first idea to store the token addresses in the `Pair` struct but I decided not to since `(token_0, token_1)` are used as the key in the mapping where the `Pair`s are stored in the contract, meaning that the tokens must be valid - the `Pair` is used only internally in the contract code. I should have left a comment about it though...

If we didn't have to store the `Pair`s by `(token_0, token_1)` then the check would be needed.